### PR TITLE
Questionnaire

### PR DESCRIPTION
**Pull request by carmeltsadok**

### CODE REVIEW

File: shared/question/question.ts [hunk 1]
```diff
@@ -0,0 +1,274 @@
+interface ScaleLimit {
+    tag: string;
+    value: number;
+}
+
+enum QuestionTypes {
```
Comment (yonatandt) [inline shared/question/question.ts:6]: `QuestionType` instead of `QuestionTypes`

File: shared/question/question.ts [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+            super.isQuestionValid() &&
+            Array.isArray(this.options) &&
+            this.options.every((item) => typeof item === 'string') &&
+            !(this.hasOther !== undefined && typeof this.hasOther !== 'boolean')
+        );
+    }
+
+    isAnswerValid() {
+        if (!super.isAnswerValid()) return false;
+        const differntanswersFromOptions = this.answer.filter((checkedAnswer) => !this.options.includes(checkedAnswer));
```
Comment (yonatandt) [inline shared/question/question.ts:116]: `differntAnswersFromOptions`

Comment (yonatandt) [reply shared/question/question.ts:116]: `this.answer` may not exist

File: shared/question/question.ts [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+    }
+}
+
+const validateQuestions = (questions: [any], answerValidation: boolean) => {
+    if (!Array.isArray(questions)) {
+        return false;
+    }
+    for (let questionIdx = 0; questionIdx < questions.length; questionIdx += 1) {
+        let questionObject: any;
+        let isValid: Boolean = true;
```
Comment (yonatandt) [inline shared/question/question.ts:221]: `isValid` is not needed

File: shared/authWrapper.ts [hunk 4]
```diff
@@ -40,13 +80,14 @@ const handleUserToken = (token: string): User | null => {
  * @returns a function that wraps func with authentication handling that either send
  * the authenticated user to func or returns a Unauthenticated error to the client.
  */
-export default (func: AzureFunction): AzureFunction => {
+export const authWrapper = (func: AzureFunction): AzureFunction => {
     return async (context: Context, req: HttpRequest): Promise<HttpTriggerResult> => {
         const header = req.headers?.authorization;
+        const requesterType = req.headers && req.headers['requester-type'];
         if (header && header.split(' ')[1]) {
```
Comment (yonatandt) [inline shared/authWrapper.ts:87]: add a check on `requesterType`

File: AddQuestionnaireToEventResults/index.ts [hunk 5]
```diff
@@ -0,0 +1,10 @@
+import { AzureFunction, Context } from '@azure/functions';
+import EventModel from '../shared/event/event.model';
+import getConnection from '../shared/mongoose';
+
+const AddQuestionnaireToEventResults: AzureFunction = async (_context: Context, message: any): Promise<void> => {
+    await getConnection();
+    await EventModel.findByIdAndUpdate(message.eventId, { $push: { 'results.questionnaire': message.questionnaireInstanceId } }, { new: true });
+};
+
```
Comment (vitaly-furman) [inline AddQuestionnaireToEventResults/index.ts:9]: Add try and catch for exception handling

File: CreateNode/index.ts [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+            throw new ValidationFunctionError(functionName, "must specfiy unitId for node of type 'Interview'");
+        }
+
+        await getConnection();
+
+        const node: Node = await NodeModel.create(req.body);
+
+        return {
+            status: 201,
+            body: node,
```
Comment (vitaly-furman) [inline CreateNode/index.ts:33]: Store the status code in env variable

File: CreateQuestionnaireSchema/index.ts [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+        if (validationError) {
+            throw new ValidationFunctionError(functionName, validationError.message);
+        }
+        if (!validateQuestions(req.body.questions, false)) {
+            throw new ValidationFunctionError(functionName, 'The questions are invalid');
+        }
+        await getConnection();
+        const creationReq = { ...req.body };
+        creationReq.createdBy = user.id;
+        creationReq.updatedBy = user.id;
```
Comment (vitaly-furman) [inline CreateQuestionnaireSchema/index.ts:30]: Can be added to single line
const creationReq = { ...req.body, createdBy: user.id,  updatedBy = user.id};

File: CreateQuestionnaireSchema/index.ts [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+
+const createQuestionnaireSchema: AzureFunction = async (_context: Context, req: HttpRequest, user: User): Promise<HttpTriggerResult> => {
+    try {
+        if (user.role !== Role.Mada) {
+            throw new ForbiddenFunctionError(functionName, user);
+        }
+        const validationError: ValidationError = questionnaireSchemaValidation.validate(req).error;
+        if (validationError) {
+            throw new ValidationFunctionError(functionName, validationError.message);
+        }
```
Comment (vitaly-furman) [inline CreateQuestionnaireSchema/index.ts:23]: Better use object destructuring like this:
const { error } = questionnaireSchemaValidation.validate(req)
if(error) ...

Comment (lustigy96) [reply CreateQuestionnaireSchema/index.ts:23]: Cool, but this is how all functions are written and I think that its better when all the code is unified

**Pull request merged**